server/data/data-cassandra/src/main/java/org/apache/james/sieve/cassandra/CassandraSieveRepository.java

```diff
@@ -73,16 +72,29 @@ public ZonedDateTime getActivationDateForActiveScript(User user) throws ScriptNo
 
     @Override
     public void haveSpace(User user, ScriptName name, long newSize) throws QuotaExceededException {
-        throwOnOverQuota(user, spaceThatWillBeUsedByNewScript(user, name, newSize));
+        reThrowQuotaExceededException(() ->
+            spaceThatWillBeUsedByNewScript(user, name, newSize)
+                .flatMap(sizeDifference -> throwOnOverQuota(user, sizeDifference))
+                .block());
     }
 
-    private void throwOnOverQuota(User user, Mono<Long> sizeDifference) throws QuotaExceededException {
-        CompletableFuture<Optional<QuotaSize>> userQuotaFuture = cassandraSieveQuotaDAO.getQuota(user);
-        CompletableFuture<Optional<QuotaSize>> globalQuotaFuture = cassandraSieveQuotaDAO.getQuota();
-        CompletableFuture<Long> spaceUsedFuture = cassandraSieveQuotaDAO.spaceUsedBy(user);
+    private Mono<Void> throwOnOverQuota(User user, Long sizeDifference) {
+        Mono<Optional<QuotaSize>> userQuotaMono = cassandraSieveQuotaDAO.getQuota(user);
+        Mono<Optional<QuotaSize>> globalQuotaMono = cassandraSieveQuotaDAO.getQuota();
+        Mono<Long> spaceUsedMono = cassandraSieveQuotaDAO.spaceUsedBy(user);
 
-        new SieveQuota(spaceUsedFuture.join(), limitToUse(userQuotaFuture, globalQuotaFuture))
-            .checkOverQuotaUponModification(sizeDifference.block());
+        return limitToUse(userQuotaMono, globalQuotaMono).zipWith(spaceUsedMono)
+            .flatMap(pair -> checkOverQuotaUponModification(sizeDifference, pair.getT2(), pair.getT1()));
+    }
+
+    private Mono<Void> checkOverQuotaUponModification(Long sizeDifference, Long spaceUsed, Optional<QuotaSize> limit) {
+        try {
+            new SieveQuota(spaceUsed, limit)
+                    .checkOverQuotaUponModification(sizeDifference);
+            return Mono.empty();
+        } catch (QuotaExceededException e) {
```

Comment:
Can probably avoid the try catch with Mono error handling here too

Reply:
Not here, checked exceptions are wrapped by Reactor, which breaks the contract.